internal/provider/endpoint/endpoint.go

```diff
@@ -0,0 +1,216 @@
+		err = zoneAccessor.EnsureEndpoint(updateEndpoint)
+		if err != nil {
+			me = multierror.Append(me, err)
+		}
+	}
+	for _, deleteEndpoint := range changes.Delete {
+		err = zoneAccessor.RemoveEndpoint(deleteEndpoint)
+		if err != nil {
+			me = multierror.Append(me, err)
+		}
+	}
+	_, err = e.client.Resource(e.gvr).Namespace(e.providerNS).Update(ctx, zoneAccessor.GetObject(), metav1.UpdateOptions{})
+	if err != nil {
+		me = multierror.Append(me, err)
+	}
+	return me.ErrorOrNil()
+}
+
+// Records returns the list of endpoints
+func (e *EndpointProvider) Records(ctx context.Context) ([]*endpoint.Endpoint, error) {
+	zoneAccessor, err := e.getZoneAccessorForZoneIDFilter(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	return zoneAccessor.GetEndpoints(), nil
+}
+
+func (e *EndpointProvider) DNSZones(ctx context.Context) ([]dnsOperatorProvider.DNSZone, error) {
+	var hzs []dnsOperatorProvider.DNSZone
+	zones, err := e.client.Resource(e.gvr).Namespace(e.providerNS).List(
+		ctx,
+		metav1.ListOptions{LabelSelector: e.labelSelector},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, z := range zones.Items {
+		za, err := NewEndpointAccessor(&z)
+		if err != nil {
+			e.logger.Info("badly formatted zone", "zone name", z.GetName())
+			continue
+		}
+		hz := dnsOperatorProvider.DNSZone{
+			ID:      z.GetName(),
+			DNSName: za.GetRootHost(),
+		}
+		hzs = append(hzs, hz)
+	}
+	return hzs, nil
+}
+
+func (e *EndpointProvider) ProviderSpecific() dnsOperatorProvider.ProviderSpecificLabels {
+	return dnsOperatorProvider.ProviderSpecificLabels{}
+}
+
+func (e *EndpointProvider) getZoneAccessorForZoneIDFilter(ctx context.Context) (*endpointAccessor, error) {
+	if !e.config.ZoneIDFilter.IsConfigured() {
+		return nil, fmt.Errorf("no zone id filter specified for Endpoint Provider")
+	}
+
+	e.logger.Info("getting zone accessor for zone id filter", "filter", e.config.ZoneIDFilter, "object", e.object)
+
+	zone, err := e.client.Resource(e.gvr).Namespace(e.object.GetNamespace()).Get(ctx, e.config.ZoneIDFilter.ZoneIDs[0], metav1.GetOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return NewEndpointAccessor(zone)
+}
+
+// GetDomainFilter implements provider.Provider.
+// Subtle: this method shadows the method (BaseProvider).GetDomainFilter of EndpointProvider.BaseProvider.
+func (e *EndpointProvider) GetDomainFilter() endpoint.DomainFilter {
+	return endpoint.DomainFilter{}
+}
+
+// Name implements provider.Provider.
+func (e *EndpointProvider) Name() dnsOperatorProvider.DNSProviderName {
+	return dnsOperatorProvider.DNSProviderEndpoint
+}
+
+// AdjustEndpoints nothing to do here
+func (e *EndpointProvider) AdjustEndpoints(endpoints []*endpoint.Endpoint) ([]*endpoint.Endpoint, error) {
+	return endpoints, nil
+}
+
+func NewProviderFromSecret(ctx context.Context, client dynamic.Interface, s *v1.Secret, providerConfig dnsOperatorProvider.Config) (dnsOperatorProvider.Provider, error) {
+	logger := log.FromContext(ctx).WithName("endpoint-dns")
```

Comment:
Should this add checks to make sure the secret exists and has a namespace? This will be important for this provider:

```go
if s == nil {
    return nil, fmt.Errorf("secret cannot be nil")
}

if s.GetNamespace() == "" {
    return nil, fmt.Errorf("namespace not set")
}
```